clang/lib/CodeGen/CGDecl.cpp

```diff
@@ -2120,27 +2066,33 @@ void CodeGenFunction::EmitAutoVarCleanups(const AutoVarEmission &emission) {
   // us from jumping into any of these scopes anyway.
   if (!HaveInsertPoint()) return;
 
-  const VarDecl &D = *emission.Variable;
+  // If we're initializing directly on the heap, _Block_object_destroy will
+  // handle destruction, so we don't need to perform cleanups directly.
```

Comment:
I think the right way to think of this is that the stack copy of the byref is always uninitialized in this situation, so there's nothing for us to destroy.  Basically, we make an uninitialized byref structure on the stack, ask the runtime to move it to the heap with a trivial move function, and then initialize the object on the heap directly.

Reply:
Indeed... and looking at this again, I put too much code inside the if statement.

Not only that, there's an existing bug here when combining `__attribute__((cleanup))` with `__block`: if the variable was copied to the heap, the cleanup function will be called after the byref allocation has already been freed.

Will fix.